src/analysis/normed_space/normed_group_hom.lean

```diff
@@ -0,0 +1,465 @@
+  (λ hf, le_antisymm (real.Inf_le _ bounds_bdd_below
+    ⟨ge_of_eq rfl, λ _, le_of_eq (by { rw [zero_mul, hf], exact norm_zero })⟩)
+    (op_norm_nonneg _))
+
+-- see Note [addition on function coercions]
+@[simp] lemma coe_zero : ⇑(0 : normed_group_hom V₁ V₂) = (0 : V₁ → V₂) := rfl
+@[simp] lemma zero_apply (v : V₁) : (0 : normed_group_hom V₁ V₂) v = 0 := rfl
+
+variables {f g}
+
+/-! ### The identity normed group hom -/
+
+/-- The identity as a continuous normed group hom. -/
+@[simps]
+def id : normed_group_hom V V :=
+(add_monoid_hom.id V).mk_continuous 1 (by simp [le_refl])
+
+/-- The norm of the identity is at most `1`. It is in fact `1`, except when the space is trivial
+where it is `0`. It means that one can not do better than an inequality in general. -/
+lemma norm_id_le : ∥(id : normed_group_hom V V)∥ ≤ 1 :=
+op_norm_le_bound _ zero_le_one (λx, by simp)
+
+/-- If a space is non-trivial, then the norm of the identity equals `1`. -/
+lemma norm_id [nontrivial V] : ∥(id : normed_group_hom V V)∥ = 1 :=
+le_antisymm norm_id_le $ let ⟨x, hx⟩ := exists_ne (0 : V) in
+have _ := (id : normed_group_hom V V).ratio_le_op_norm x,
+by rwa [id_apply, div_self (ne_of_gt $ norm_pos_iff.2 hx)] at this
+
+/-! ### The negation of a normed group hom -/
+
+/-- Opposite of a normed group hom. -/
+instance : has_neg (normed_group_hom V₁ V₂) :=
+⟨λ f, (-f.to_add_monoid_hom).mk_continuous (∥f∥) (λ v, by simp [le_op_norm f v])⟩
+
+-- see Note [addition on function coercions]
+@[simp] lemma coe_neg (f : normed_group_hom V₁ V₂) : ⇑(-f) = (-f : V₁ → V₂) := rfl
+@[simp] lemma neg_apply (f : normed_group_hom V₁ V₂) (v : V₁) :
+  (-f : normed_group_hom V₁ V₂) v = - (f v) := rfl
+
+lemma op_norm_neg (f : normed_group_hom V₁ V₂) : ∥-f∥ = ∥f∥ :=
+by simp only [norm_def, coe_neg, norm_neg, pi.neg_apply]
+
+/-! ### Subtraction of normed group homs -/
+
+/-- Subtraction of normed group homs. -/
+instance : has_sub (normed_group_hom V₁ V₂) :=
+⟨λ f g,
+{ bound' :=
+  begin
+    simp only [add_monoid_hom.sub_apply, add_monoid_hom.to_fun_eq_coe, sub_eq_add_neg],
+    exact (f + -g).bound'
+  end,
+  .. (f.to_add_monoid_hom - g.to_add_monoid_hom) }⟩
+
+-- see Note [addition on function coercions]
+@[simp] lemma coe_sub (f g : normed_group_hom V₁ V₂) : ⇑(f - g) = (f - g : V₁ → V₂) := rfl
+@[simp] lemma sub_apply (f g : normed_group_hom V₁ V₂) (v : V₁) :
+  (f - g : normed_group_hom V₁ V₂) v = f v - g v := rfl
+
+/-! ### Normed group structure on normed group homs -/
+
+/-- Homs between two given normed groups form a commutative additive group. -/
+instance : add_comm_group (normed_group_hom V₁ V₂) :=
+{ add := (+), zero := 0, neg := has_neg.neg, sub := has_sub.sub,
+  add_assoc := by { intros, ext, simp [add_assoc] },
+  add_comm := by { intros, ext, simp [add_comm] },
+  zero_add := by { intros, ext, simp },
+  add_zero := by { intros, ext, simp },
+  sub_eq_add_neg := by { intros, ext, simp [sub_eq_add_neg] },
+  add_left_neg := by { intros, ext, simp } }
```

Comment:
untested, but something like this will be faster:
```suggestion
function.injective.add_comm_group_sub coe_fn coe_inj rfl (λ _ _, rfl) (λ _, rfl) (λ _ _, rfl)
```